### PR TITLE
Adding Kubernetes native secrets to secret stores

### LINKED ISSUE #5520: Design: extend the use-cases of Applications.Core/secretStores
## Overview of feature request

[TLS termination specification](https://github.com/project-radius/radius/issues/2310#issuecomment-1516802963) proposes the use-case of gateway for Applications.Core/secretStores. Beyond this use-case, we want to extend the capability of secretstores to the other radius resources, such as `Applications.Core/containers`. 

This issue is to investigate the other scenarios and use-cases for non-gateway resources.

[AB#7794](https://dev.azure.com/azure-octo/e61041b4-555f-47ae-95b2-4f8ab480ea57/_workitems/edit/7794)

### PR DESCRIPTION
**Pull request by superbeeny**
# Description
Adding secret store capabilities to reference kubernetes secrets natively

## Type of change

- This pull request adds or changes features of Radius and has an approved issue (#5520) .

Fixes: #5520

### CODE REVIEW

**Comment by rynowak** [comment]
Thanks for submitting this @superbeeny ! This is a substantial feature that has been on the backlog for a long time.

I'll start reviewing the code. While we're making progress on that, a few month things we'll need to do.

- [ ] Post and get signoff on a feature design spec at radius-project/design-notes. We require a design doc for API changes. @willtsai is on the hook for this.
- [ ] Post and get signoff on a an engineering design doc at radius-project/design-notes. We require a design doc for API changes.
- [ ] Update existing functional tests and samples to handle the breaking change.
- [ ] Update docs (post-merge) to make sure they are valid.

/cc @Reshrahim as FYI

File: typespec/Applications.Core/containers.tsp [hunk 1]
```diff
@@ -240,7 +240,7 @@ model Container {
   imagePullPolicy?: ImagePullPolicy;
 
   @doc("environment")
-  env?: Record<string>;
+  env?: Record<EnvironmentVariable>;
```
Comment (rynowak) [inline typespec/Applications.Core/containers.tsp:243]: Note to reviewers, we have to make a breaking change to add this feature. We'll discuss the API changes in design review.

Comment (rynowak) [reply typespec/Applications.Core/containers.tsp:243]: Example of the updated usage here: https://github.com/radius-project/radius/pull/7744/files#diff-19f5067c27bb52be17b49eb73842441f50662b7ef8317d161ed0e6863a26f0d0R29

File: pkg/corerp/api/v20231001preview/container_conversion.go [hunk 2]
```diff
@@ -151,6 +156,64 @@ func (src *ContainerResource) ConvertTo() (v1.DataModelInterface, error) {
 	return converted, nil
 }
 
+// toEnvDataModel: Converts from versioned datamodel to base datamodel
+func toEnvDataModel(e map[string]*EnvironmentVariable) (map[string]datamodel.EnvironmentVariable, error) {
```
Comment (rynowak) [inline pkg/corerp/api/v20231001preview/container_conversion.go:160]: @lakshmimsft @ytimocin - can one of you help review the converter code?

File: pkg/corerp/renderers/container/render.go [hunk 3]
```diff
@@ -109,6 +109,20 @@ func (r Renderer) GetDependencyIDs(ctx context.Context, dm v1.DataModelInterface
 		}
 	}
 
+	for _, envvars := range properties.Container.Env {
```
Comment (rynowak) [inline pkg/corerp/renderers/container/render.go:112]: Nit: I'd suggest a comment here describing why we need to do all of this. Something like: `Environment variables might reference a `secretStores` resource, so walk through them and find any references`.

File: pkg/corerp/renderers/container/render_test.go [hunk 4]
```diff
@@ -173,6 +177,27 @@ func Test_GetDependencyIDs_Success(t *testing.T) {
...
+				},
+				envVarName3: {
+					ValueFrom: &datamodel.EnvironmentVariableReference{
+						SecretRef: &datamodel.EnvironmentVariableSecretReference{
+							Source: envVarSource3,
+							Key:    envVarValue3,
+						},
+					},
+				},
+			},
```
Comment (rynowak) [inline pkg/corerp/renderers/container/render_test.go:200]: Could you add something to cover the namespace-qualified kubernetes-secret case?

Looks like it's not covered: https://app.codecov.io/gh/radius-project/radius/pull/7744?src=pr&el=tree&filepath=pkg%2Fcorerp%2Fapi%2Fv20231001preview%2Fcontainer_conversion.go&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=radius-project#bc8195f5174275625d34d36017477e84-R654

File: pkg/corerp/api/v20231001preview/container_conversion.go [hunk 5]
```diff
@@ -101,6 +102,10 @@ func (src *ContainerResource) ConvertTo() (v1.DataModelInterface, error) {
 		}
 	}
 
+	convertedEnvironmentVariables, err := toEnvDataModel(src.Properties.Container.Env)
+	if err != nil {
+		return nil, err
+	}
```
Comment (ytimocin) [inline pkg/corerp/api/v20231001preview/container_conversion.go:108]: nit: you can add a new line after this.

File: pkg/corerp/datamodel/container.go [hunk 6]
```diff
@@ -116,16 +116,33 @@ type ConnectionProperties struct {
...
+	Volumes         map[string]VolumeProperties    `json:"volumes,omitempty"`
+	Command         []string                       `json:"command,omitempty"`
+	Args            []string                       `json:"args,omitempty"`
+	WorkingDir      string                         `json:"workingDir,omitempty"`
+}
+
+// EnvironmentVariable - Environment variable for the container
+type EnvironmentVariable struct {
+	Value     *string                       `json:"value,omitempty"`
+	ValueFrom *EnvironmentVariableReference `json:"valueFrom,omitempty"`
```
Comment (ytimocin) [inline pkg/corerp/datamodel/container.go:134]: Can we also add documentation for the properties? Maybe with some examples?

File: pkg/corerp/renderers/container/render.go (hunk 3 above)
Comment (ytimocin) [inline pkg/corerp/renderers/container/render.go:112]: Should we do this camel case? Like **envVars**?

File: pkg/corerp/renderers/container/render.go [hunk 7]
```diff
@@ -109,6 +109,20 @@ func (r Renderer) GetDependencyIDs(ctx context.Context, dm v1.DataModelInterface
...
+		if envvars.ValueFrom != nil && envvars.ValueFrom.SecretRef != nil {
+			if strings.HasPrefix(envvars.ValueFrom.SecretRef.Source, "/") {
```
Comment (ytimocin) [inline pkg/corerp/renderers/container/render.go:114]: Does this mean that if the **Source** has the prefix "/" it could be a resource id? If there is a secret source that doesn't have this prefix, should we log this or error out?

File: pkg/corerp/renderers/container/render.go [hunk 8]
```diff
@@ -109,6 +109,20 @@ func (r Renderer) GetDependencyIDs(ctx context.Context, dm v1.DataModelInterface
...
+				resourceID, err := resources.ParseResource(envvars.ValueFrom.SecretRef.Source)
+				if err != nil {
+					return nil, nil, v1.NewClientErrInvalidRequest(fmt.Sprintf("invalid source: %s. Must be either a kubernetes secret name or a valid resourceID", envvars.ValueFrom.SecretRef.Source))
+				}
+				if resources_radius.IsRadiusResource(resourceID) {
```
Comment (ytimocin) [inline pkg/corerp/renderers/container/render.go:119]: If it is not a Radius resource, what do we do? Do we just ignore it?

File: pkg/corerp/renderers/container/render.go [hunk 9]
```diff
@@ -616,6 +633,58 @@ func (r Renderer) makeDeployment(
 	return outputResources, secretData, nil
 }
 
+// Function to convert from map[string]EnvironmentVariable to map[string]corev1.EnvVar
+func convertEnvVar(key string, env datamodel.EnvironmentVariable, options renderers.RenderOptions) (corev1.EnvVar, error) {
```
Comment (ytimocin) [inline pkg/corerp/renderers/container/render.go:637]: The documentation should start with the function name.

File: pkg/corerp/renderers/container/render.go [hunk 10]
```diff
@@ -616,6 +633,58 @@ func (r Renderer) makeDeployment(
...
+	if env.Value != nil {
+		return corev1.EnvVar{Name: key, Value: *env.Value}, nil
+	} else {
+		// There are two cases to handle here:
+		// 1. The value comes from a kubernetes secret
+		// 2. The value comes from a radius resource id.
+
+		// If the value comes from a kubernetes secret, we'll reference it.
+		if strings.HasPrefix(env.ValueFrom.SecretRef.Source, "/") {
```
Comment (ytimocin) [inline pkg/corerp/renderers/container/render.go:646]: Does having the prefix "/" mean that it is a Kubernetes secret? Should we have a separate function for this if we are going to use it in a few different places?

File: pkg/corerp/renderers/container/render.go [hunk 11]
```diff
@@ -616,6 +633,58 @@ func (r Renderer) makeDeployment(
...
+				},
+			}, nil
+
+		} else {
+			return corev1.EnvVar{
+				Name: key,
+				ValueFrom: &corev1.EnvVarSource{
+					SecretKeyRef: &corev1.SecretKeySelector{
+						LocalObjectReference: corev1.LocalObjectReference{
+							Name: env.ValueFrom.SecretRef.Source,
```
Comment (ytimocin) [inline pkg/corerp/renderers/container/render.go:677]: I think we can extract the name decision to a separate function. We can use it in other places too.

File: pkg/to/value.go [hunk 12]
```diff
@@ -28,6 +28,11 @@ func String(s *string) string {
 	return ""
 }
 
+// StringPtr returns a pointer to the string passed in.
+func StringPtr(s string) *string {
```
Comment (ytimocin) [inline pkg/to/value.go:32]: Thanks for this addition.

File: swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json [hunk 13]
```diff
@@ -3320,6 +3320,89 @@
         }
       }
     },
+    "EnvironmentVariable": {
+      "type": "object",
+      "description": "Envinronment variables type",
```
Comment (ytimocin) [inline swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json:3325]:
```suggestion
      "description": "Environment variable type",
```

File: test/functional-portable/corerp/noncloud/resources/container_test.go [hunk 14]
```diff
@@ -354,3 +354,29 @@ func Test_Container_FailDueToBadHealthProbe(t *testing.T) {
...
+			RPResources: &validation.RPResourceSet{
+				Resources: []validation.RPResource{},
+			},
+			K8sObjects: &validation.K8sObjectSet{
+				Namespaces: map[string][]validation.K8sObject{
+					appNamespace: {
+						validation.NewK8sPodForResource(name, "cntr-cntr-secrets"),
+					},
+				},
+			},
```
Comment (ytimocin) [inline test/functional-portable/corerp/noncloud/resources/container_test.go:377]: Do we need to add **PostStepVerify** here and get the secret to make sure that it is set?

File: typespec/Applications.Core/containers.tsp [hunk 15]
```diff
@@ -264,6 +264,30 @@ model Container {
   workingDir?: string;
 }
 
+@doc("Envinronment variables type")
```
Comment (ytimocin) [inline typespec/Applications.Core/containers.tsp:267]:
```suggestion
@doc("Environment variables type")
```

File: pkg/corerp/api/v20231001preview/container_conversion.go (hunk 2 above)
Comment (ytimocin) [inline pkg/corerp/api/v20231001preview/container_conversion.go:160]: We can rename this to **toEnvironmentVariableDataModel**.

File: pkg/corerp/api/v20231001preview/container_conversion.go [hunk 16]
```diff
@@ -151,6 +156,64 @@ func (src *ContainerResource) ConvertTo() (v1.DataModelInterface, error) {
...
+
+	m := map[string]datamodel.EnvironmentVariable{}
+
+	for key, val := range e {
+		if val == nil {
+			return nil, v1.NewClientErrInvalidRequest(fmt.Sprintf("Environment variable %s is nil", key))
+		}
+		if val.Value != nil && val.ValueFrom != nil {
+			return nil, v1.NewClientErrInvalidRequest(fmt.Sprintf("Environment variable %s has both value and secret value", key))
+		}
```
Comment (ytimocin) [inline pkg/corerp/api/v20231001preview/container_conversion.go:170]: Could you add some information about **ValueFrom** and **Value** here?

**Review by ytimocin — COMMENTED** [review]
This is looking great! Added some comments and asked some questions. Thanks a lot!

File: pkg/corerp/datamodel/container.go (hunk 6 above)
Comment (superbeeny) [reply pkg/corerp/datamodel/container.go:134]: Do you mean in the code or in the documentation?

**Comment by superbeeny** [comment]
DCO not my fault ;)

File: pkg/corerp/api/v20231001preview/container_conversion.go (hunk 2 above)
Comment (superbeeny) [reply pkg/corerp/api/v20231001preview/container_conversion.go]: Also renamed the fromEnvVarDataModel to be consistent

File: test/functional-portable/corerp/noncloud/resources/container_test.go (hunk 14 above)
Comment (superbeeny) [reply test/functional-portable/corerp/noncloud/resources/container_test.go:377]: @rynowak and I discussed this at the time and decided not to, no problem in revisiting that choice here.

File: pkg/to/value.go (hunk 12 above)
Comment (superbeeny) [reply pkg/to/value.go:32]: Added a test for this function too

**Comment by superbeeny** [comment]
Functional Test - https://github.com/radius-project/radius/actions/runs/9997221043/job/27633433901#step:27:216

failed with a 403 error which I don't think is related to this PR

File: pkg/to/value.go (hunk 12 above)
Comment (lakshmimsft) [reply pkg/to/value.go:32]: @ytimocin, @rynowak perhaps you can confirm but looking at functions in this package, seems like it is intended to work with input pointers and convert to their respective values and not vice versa. 
If that's the case, I would suggest using existing function to.Ptr() instead defined [here](https://github.com/radius-project/radius/blob/53e19b23e0736161da7804de0a297cefe13f6a71/pkg/to/ptr.go#L22)

File: pkg/corerp/api/v20231001preview/container_conversion.go [hunk 17]
```diff
@@ -151,6 +157,65 @@ func (src *ContainerResource) ConvertTo() (v1.DataModelInterface, error) {
 	return converted, nil
 }
 
+// toEnvDataModel: Converts from versioned datamodel to base datamodel
```
Comment (lakshmimsft) [inline pkg/corerp/api/v20231001preview/container_conversion.go:160]: nit: // toEnvironmentVariableDataModel:

File: pkg/corerp/api/v20231001preview/container_conversion.go [hunk 18]
```diff
@@ -151,6 +157,65 @@ func (src *ContainerResource) ConvertTo() (v1.DataModelInterface, error) {
...
+				},
+			}
+
+		}
+
+	}
+	return m, nil
+}
+
+// fromEnvDataModel: Converts from base datamodel to versioned datamodel
```
Comment (lakshmimsft) [inline pkg/corerp/api/v20231001preview/container_conversion.go:194]: nit: I think we may want to keep the same name format. If it's  toEnvironmentVariableDataModel() then a corresponding fromEnvironmentVariableDataModel()

File: pkg/corerp/api/v20231001preview/container_conversion.go (hunk 17 above)
Comment (superbeeny) [reply pkg/corerp/api/v20231001preview/container_conversion.go]: Fixed as requested

File: pkg/corerp/api/v20231001preview/container_conversion.go (hunk 16 above)
Comment (superbeeny) [reply pkg/corerp/api/v20231001preview/container_conversion.go:172]: Added comment as requested

File: typespec/Applications.Core/containers.tsp [hunk 19]
```diff
@@ -264,6 +264,30 @@ model Container {
...
+  valueFrom?: EnvironmentVariableReference;
+}
+
+@doc("The reference to the variable")
+model EnvironmentVariableReference {
+  @doc("The secret reference")
+  secretRef: EnvironmentVariableSecretReference;
+}
+
+@doc("The secret reference")
```
Comment (lakshmimsft) [inline typespec/Applications.Core/containers.tsp:282]: Suggestion: We might be able to reuse an existing model SecretReference from Applications.Core with the same structure : https://github.com/radius-project/radius/blob/46bf036b9d3c4d18881a568845d71cf382d8f701/typespec/Applications.Core/environments.tsp#L177-L184

File: pkg/corerp/datamodel/container.go (hunk 6 above)
Comment (ytimocin) [reply pkg/corerp/datamodel/container.go:134]: kind of like:

```suggestion
	// Value is the property for ... Example is ...
	Value     *string                       `json:"value,omitempty"`
	// ValueFrom is the property for ... Example is ...
	ValueFrom *EnvironmentVariableReference `json:"valueFrom,omitempty"`
```

File: pkg/corerp/api/v20231001preview/container_conversion.go (hunk 16 above)
Comment (ytimocin) [reply pkg/corerp/api/v20231001preview/container_conversion.go:172]: Is the **valueFrom** the one from here: https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#use-configmap-defined-environment-variables-in-pod-commands?

File: pkg/to/value.go (hunk 12 above)
Comment (ytimocin) [reply pkg/to/value.go:32]: I agree with @lakshmimsft here. I don't think we need this function. You can see the usages of **to.Ptr** here: https://github.com/search?q=repo%3Aradius-project%2Fradius%20%22to.Ptr%22&type=code.

File: test/functional-portable/corerp/noncloud/resources/container_test.go (hunk 14 above)
Comment (ytimocin) [reply test/functional-portable/corerp/noncloud/resources/container_test.go:377]: sounds good to me

File: pkg/corerp/renderers/container/render.go (hunk 7 above)
Comment (superbeeny) [reply pkg/corerp/renderers/container/render.go]: Source is either the resource id of a radius resource or a kubernetes secret reference.

File: pkg/corerp/renderers/container/render.go (hunk 10 above)
Comment (superbeeny) [reply pkg/corerp/renderers/container/render.go:647]: If we do end up using it in another function then that may be the right time to extract it to its own function.

File: pkg/corerp/api/v20231001preview/container_conversion.go [hunk 20]
```diff
@@ -151,6 +157,65 @@ func (src *ContainerResource) ConvertTo() (v1.DataModelInterface, error) {
 	return converted, nil
 }
 
+// toEnvironmentVariableDataModel: Converts from versioned datamodel to base datamodel
+func toEnvironmentVariableDataModel(e map[string]*EnvironmentVariable) (map[string]datamodel.EnvironmentVariable, error) {
+
```
Comment (ytimocin) [inline pkg/corerp/api/v20231001preview/container_conversion.go:162]: nit: extra line can be removed

File: pkg/corerp/datamodel/container.go [hunk 21]
```diff
@@ -116,16 +116,36 @@ type ConnectionProperties struct {
...
+// EnvironmentVariableReference - Environment variable reference for the container
+type EnvironmentVariableReference struct {
+	SecretRef *EnvironmentVariableSecretReference `json:"secretRef"`
+}
+
+// EnvironmentVariableSecretReference - Environment variable secret reference for the container
+type EnvironmentVariableSecretReference struct {
+	// Source is either the resource id of a radius resource or a kubernetes secret reference.
+	Source string `json:"source"`
+	Key    string `json:"key"`
```
Comment (ytimocin) [inline pkg/corerp/datamodel/container.go:148]: nit: add doc

Comment (rynowak) [reply pkg/corerp/datamodel/container.go:148]: Yep.

File: pkg/kubeutil/client.go [hunk 22]
```diff
@@ -28,6 +28,9 @@ import (
 	"k8s.io/client-go/rest"
 	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
 	csidriver "sigs.k8s.io/secrets-store-csi-driver/apis/v1alpha1"
+
+	// Import kubernetes auth plugins
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
```
Comment (rynowak) [inline pkg/kubeutil/client.go:33]: Note: @superbeeny ran into an issue where UCP isn't doing auth to Kubernetes correctly for OIDC. This is the fix.

File: test/functional-portable/cli/noncloud/cli_test.go [hunk 23]
```diff
@@ -69,7 +69,8 @@ func verifyRecipeCLI(ctx context.Context, t *testing.T, test rp.RPTest) {
 	resourceType := "Applications.Datastores/redisCaches"
 	file := "../../../testrecipes/test-bicep-recipes/corerp-redis-recipe.bicep"
 
-	target := fmt.Sprintf("br:%s/dev/test-bicep-recipes/redis-recipe:%s", registry, generateUniqueTag())
+	target := fmt.Sprintf("br:%s/dev/test-bicep-recipes/redis-recipe:%s",
+		strings.TrimPrefix(registry, "registry="), generateUniqueTag())
```
Comment (rynowak) [inline test/functional-portable/cli/noncloud/cli_test.go:73]: Context for this change? Was this broken?

File: versions.yaml [hunk 24]
```diff
@@ -2,8 +2,7 @@ supported:
   - channel: '0.37'
     version: 'v0.37.0-rc1'
   - channel: '0.36'
-    version: 'v0.36.0'
-deprecated:
+    version: 'v0.36.0-rc1'
```
Comment (rynowak) [inline versions.yaml:5]: Can you revert the changes to this file? Looks like a bad merge.

For context this file drives our release process. It shouldn't change in this PR.

**Review by rynowak — COMMENTED** [review]
This looks great to me overall. Please resolve the outstanding questions/feedback and lets figure out what's causing the test failures. 

/cc @ytimocin @kachawla any remaining concerns?

**Comment by rynowak** [comment]
Looks like there's one more file to update: https://github.com/radius-project/radius/blob/main/test/functional-portable/corerp/cloud/resources/testdata/corerp-resources-container-workload.bicep#L56

I think this was added while your PR was in progress.

File: pkg/corerp/renderers/container/render.go (hunk 8 above)
Comment (kachawla) [reply pkg/corerp/renderers/container/render.go:120]: +1 shouldn't it error out if it happens?

File: pkg/corerp/renderers/container/render_test.go [hunk 25]
```diff
@@ -173,6 +180,35 @@ func Test_GetDependencyIDs_Success(t *testing.T) {
 					ContainerPort: 5000,
 				},
 			},
+			Env: map[string]datamodel.EnvironmentVariable{
```
Comment (kachawla) [inline pkg/corerp/renderers/container/render_test.go:183]: Do we have test coverage for error scenarios?

File: swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json [hunk 26]
```diff
@@ -3002,7 +3002,7 @@
           "type": "object",
           "description": "environment",
           "additionalProperties": {
-            "type": "string"
+            "$ref": "#/definitions/EnvironmentVariableUpdate"
```
Comment (kachawla) [inline swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json:3005]: @rynowak do you know what's making the `Update` suffix to be added in a couple of places?

File: test/functional-portable/datastoresrp/noncloud/resources/testdata/datastoresrp-resources-redis-manual.bicep [hunk 27]
```diff
@@ -1,5 +1,5 @@
 extension radius
-
+git config --global core.autocrlf true
```
Comment (kachawla) [inline test/functional-portable/datastoresrp/noncloud/resources/testdata/datastoresrp-resources-redis-manual.bicep:2]: was this intended?

Comment (superbeeny) [reply test/functional-portable/datastoresrp/noncloud/resources/testdata/datastoresrp-resources-redis-manual.bicep:2]: no :'(

File: pkg/corerp/renderers/container/render.go [hunk 28]
```diff
@@ -109,6 +109,21 @@ func (r Renderer) GetDependencyIDs(ctx context.Context, dm v1.DataModelInterface
 		}
 	}
 
+	// Environment variables can be sourced from secrets, which are resources. We need to iterate over the environment to handle any possible instances.
+	for _, envVars := range properties.Container.Env {
+		if envVars.ValueFrom != nil && envVars.ValueFrom.SecretRef != nil {
+			if strings.HasPrefix(envVars.ValueFrom.SecretRef.Source, "/") {
+				resourceID, err := resources.ParseResource(envVars.ValueFrom.SecretRef.Source)
+				if err != nil {
+					return nil, nil, v1.NewClientErrInvalidRequest(fmt.Sprintf("invalid source: %s. Must be either a kubernetes secret name or a valid resourceID", envVars.ValueFrom.SecretRef.Source))
```
[the remark this answers is not in the document]
Comment (superbeeny) [reply pkg/corerp/renderers/container/render.go:118]: If it is a kubernetes secret name it won't reach this part of the code as the format would be <namespace>/<secret> or just <secret>

File: pkg/corerp/renderers/container/render.go [hunk 29]
```diff
@@ -616,6 +634,64 @@ func (r Renderer) makeDeployment(
...
+func convertEnvVar(key string, env datamodel.EnvironmentVariable, options renderers.RenderOptions) (corev1.EnvVar, error) {
+	if env.Value != nil {
+		return corev1.EnvVar{Name: key, Value: *env.Value}, nil
+	} else {
+		// There are two cases to handle here:
+		// 1. The value comes from a kubernetes secret
+		// 2. The value comes from a radius resource id.
+
+		// If the value comes from a kubernetes secret, we'll reference it.
+		if strings.HasPrefix(env.ValueFrom.SecretRef.Source, "/") {
```
[the remark this answers is not in the document]
Comment (superbeeny) [reply pkg/corerp/renderers/container/render.go:647]: Radius resources

File: swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json (hunk 26 above)
Comment (rynowak) [reply swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json:3005]: I'm not sure. 

I know that TSP **can** generate different schemas for `GET` and `PUT` operations. It looks like that's what's happening here, but I don't know why it's happening.

File: test/functional-portable/datastoresrp/noncloud/resources/testdata/datastoresrp-resources-redis-manual.bicep (hunk 27 above)
Comment (rynowak) [reply test/functional-portable/datastoresrp/noncloud/resources/testdata/datastoresrp-resources-redis-manual.bicep]: image

Indeed 😁

File: typespec/Applications.Core/containers.tsp (hunk 19 above)
Comment (superbeeny) [reply typespec/Applications.Core/containers.tsp]: resolved in offline discussions

File: pkg/corerp/renderers/container/render.go [hunk 30]
```diff
@@ -616,6 +634,64 @@ func (r Renderer) makeDeployment(
...
+			secretStore, ok := options.Dependencies[env.ValueFrom.SecretRef.Source].Resource.(*datamodel.SecretStore)
```
[the remark this answers is not in the document]
Comment (superbeeny) [reply pkg/corerp/renderers/container/render.go:648]: the OK would return false and the function should error out

File: pkg/corerp/api/v20231001preview/container_conversion.go [hunk 31]
```diff
@@ -151,6 +157,64 @@ func (src *ContainerResource) ConvertTo() (v1.DataModelInterface, error) {
...
+		if val.Value != nil {
+			m[key] = datamodel.EnvironmentVariable{
+				Value: val.Value,
+			}
+		} else {
+			m[key] = datamodel.EnvironmentVariable{
+				ValueFrom: &datamodel.EnvironmentVariableReference{
+					SecretRef: &datamodel.EnvironmentVariableSecretReference{
+						Source: to.String(val.ValueFrom.SecretRef.Source),
+						Key:    to.String(val.ValueFrom.SecretRef.Key),
```
[the remark this answers is not in the document]
Comment (superbeeny) [reply pkg/corerp/api/v20231001preview/container_conversion.go:182]: TSP says they are required

File: pkg/corerp/api/v20231001preview/container_conversion.go [hunk 32]
```diff
@@ -151,6 +157,64 @@ func (src *ContainerResource) ConvertTo() (v1.DataModelInterface, error) {
...
+		// An environment variable can have either value(Value) or secret value(ValueFrom), but not both
+		if val.Value != nil && val.ValueFrom != nil {
+			return nil, v1.NewClientErrInvalidRequest(fmt.Sprintf("Environment variable %s has both value and secret value", key))
+		}
+
+		if val.Value != nil {
+			m[key] = datamodel.EnvironmentVariable{
+				Value: val.Value,
+			}
+		} else {
```
[the remark this answers is not in the document]
Comment (superbeeny) [reply pkg/corerp/api/v20231001preview/container_conversion.go:177]: Yes, updated code and added test

File: pkg/corerp/api/v20231001preview/container_conversion.go [hunk 33]
```diff
@@ -151,6 +157,64 @@ func (src *ContainerResource) ConvertTo() (v1.DataModelInterface, error) {
...
+func fromEnvironmentVariableDataModel(e map[string]datamodel.EnvironmentVariable) map[string]*EnvironmentVariable {
+	m := map[string]*EnvironmentVariable{}
+
+	for key, val := range e {
+		if val.Value != nil {
+			m[key] = &EnvironmentVariable{
+				Value: val.Value,
+			}
+		} else {
+			m[key] = &EnvironmentVariable{
```
[the remark this answers is not in the document]
Comment (superbeeny) [reply pkg/corerp/api/v20231001preview/container_conversion.go:203]: Seems you can have both, added a nil check and test

File: pkg/corerp/api/v20231001preview/container_conversion_test.go [hunk 34]
```diff
@@ -91,6 +91,22 @@ func TestContainerConvertVersionedToDataModel(t *testing.T) {
...
+					require.Equal(t, map[string]datamodel.EnvironmentVariable{
+						"DB_USER": {
+							Value: to.Ptr("DB_USER"),
+						},
+						"DB_PASSWORD": {
+							ValueFrom: &datamodel.EnvironmentVariableReference{
+								SecretRef: &datamodel.EnvironmentVariableSecretReference{
+									Source: "secret.id",
+									Key:    "DB_PASSWORD",
+								},
```
[the remark this answers is not in the document]
Comment (superbeeny) [reply pkg/corerp/api/v20231001preview/container_conversion_test.go:104]: we do now

File: pkg/corerp/renderers/container/render_test.go (hunk 25 above)
Comment (superbeeny) [reply pkg/corerp/renderers/container/render_test.go:183]: added nil checks and test

**Comment by superbeeny** [comment]
not sure how that happened

**Review by rynowak — APPROVED** [review]

**Review by kachawla — APPROVED** [review]
Thanks for addressing the comments!
